### PR TITLE
Added KubernetesStoreHandler.

### PR DESCRIPTION
**Pull request by ashcrow**
- Added KubernetesStoreHandler.
- Model now has primary_key property.
- b64.base64 now forwards encodebytes.

### CODE REVIEW

**Comment by ashcrow** [comment]
@mbarnes I have one more change for this I'm working on related to splitting the hidden attributes in `Host` in to Kubernetes secrets.

**Comment by ashcrow** [comment]
@mbarnes Feel free to start review whenever. `list` is not working correctly so I'll be fixing that later, but other than that things seem to be working fine.

**Comment by ashcrow** [comment]
Scratch the above comment. I'm refactoring a lot right now.

**Comment by ashcrow** [comment]
@mbarnes ready whenever you have time. No rush.

File: src/commissaire/handlers/hosts.py [hunk 1]
```diff
@@ -45,11 +45,13 @@ def on_get(self, req, resp):
...
             store_manager = cherrypy.engine.publish('get-store-manager')[0]
             hosts = store_manager.list(Hosts(hosts=[]))
+            print hosts
             if len(hosts.hosts) == 0:
                 raise Exception()
             resp.status = falcon.HTTP_200
             req.context['model'] = hosts
-        except:
+        except Exception as ex:
+            print ex
```
Comment (mbarnes) [inline src/commissaire/handlers/hosts.py]: Just checking -- are these `print` statements intentional or leftover debug?

Comment (ashcrow) [inline src/commissaire/handlers/hosts.py]: Leftover. I'll fixup.

File: src/commissaire/store/kubestorehandler.py [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+        certificate_path = config.get('certificate_path')
+        certificate_key_path = config.get('certificate_key_path')
+        if certificate_path and certificate_key_path:
+            self._store.cert = (certificate_path, certificate_key_path)
+
+        # TODO: Verify TLS!!!
+        self._store.verify = False
+        self._endpoint = '{0}://{1}:{2}/api/{3}'.format(
+            config['protocol'], config['host'],
+            config['port'], _API_VERSION)
```
Comment (mbarnes) [inline src/commissaire/store/kubestorehandler.py]: You have this endpoint part twice.

File: src/commissaire/store/kubestorehandler.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+
+    def _format_kwargs(self, model_instance, annotations, listing=False):
+        """
+        Formats keyword arguments used when creating a model.
+
+        :param annotations: Annotations to use when creating keyword arguments.
+        :type annotations: dict
```
Comment (mbarnes) [inline src/commissaire/store/kubestorehandler.py]: Param docs incomplete.

File: src/commissaire/store/kubestorehandler.py [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+        """
+        kwargs = {}
+        for k, v in annotations.items():
+            parts = k.split('-')
```
Comment (mbarnes) [inline src/commissaire/store/kubestorehandler.py]: I think you could use maxsplit to make this a tad simpler.
```
class_name, primary_key, model_kwarg = k.split('-', 3)
model_kwarg = model_kwarg.replace('-', '_')

File: src/commissaire/store/kubestorehandler.py [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+        return kwargs
+
+    def _format_model(self, resp_data, model_instance, listing=False):
+        """
+        Takes a model instance and figures out the proper request.
+
+        :param resp_data: Response data from the requests call.
+        :type resp_data: dict
+        :param model_instance: Model instance to save
+        :type model_instance: commissaire.model.Model
```
Comment (mbarnes) [inline src/commissaire/store/kubestorehandler.py]: Missing `:param listing:`

File: src/commissaire/store/kubestorehandler.py [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+        """
+        Creates a new instance of KubernetesStoreHandler.
+
+        :param config: Configuration details
+        :type config: dict
+        """
+        self._store = requests.Session()
+        self._endpoint = '{0}://{1}:{2}/api/{3}'.format(
+            config['protocol'], config['host'],
+            config['port'], _API_VERSION)
```
Comment (mbarnes) [inline src/commissaire/store/kubestorehandler.py]: Maybe also define `self._secrets_endpoint` for convenience below?

**Comment by mbarnes** [comment]
Nits aside, this is looking really good.

File: src/commissaire/store/kubestorehandler.py (hunk 2 above)
Comment (ashcrow) [inline src/commissaire/store/kubestorehandler.py]: Fixing.

File: src/commissaire/store/kubestorehandler.py (hunk 3 above)
Comment (ashcrow) [inline src/commissaire/store/kubestorehandler.py]: Doh! Fixing.

File: src/commissaire/store/kubestorehandler.py (hunk 5 above)
Comment (ashcrow) [inline src/commissaire/store/kubestorehandler.py]: Fixing.

File: src/commissaire/store/kubestorehandler.py (hunk 6 above)
Comment (ashcrow) [inline src/commissaire/store/kubestorehandler.py]: :+1:

File: src/commissaire/store/kubestorehandler.py (hunk 4 above)
Comment (ashcrow) [inline src/commissaire/store/kubestorehandler.py]: Good idea. Will update.

**Comment by ashcrow** [comment]
:arrow_up:

**Comment by mbarnes** [comment]
Seeing a few unit test failures...
```
======================================================================
ERROR: Verify dispatching of operations works properly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mbarnes/git/github/projectatomic/commissaire/test/test_store_kubestorehandler.py", line 75, in test__dispatch
    self.instance._save_on_namespace.assert_called_once()
  File "/home/mbarnes/commissaire/lib/python2.7/site-packages/mock/mock.py", line 721, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once

======================================================================
ERROR: Verify secret retrieval works properly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mbarnes/git/github/projectatomic/commissaire/test/test_store_kubestorehandler.py", line 115, in test__get_secret
    self.instance._store.get.assert_called_once()
  File "/home/mbarnes/commissaire/lib/python2.7/site-packages/mock/mock.py", line 721, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once

======================================================================
ERROR: Make sure we don't get empty secrets.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mbarnes/git/github/projectatomic/commissaire/test/test_store_kubestorehandler.py", line 127, in test__store_secret_without_a_valid_secret
    self.instance._store.get.assert_called_once()
  File "/home/mbarnes/commissaire/lib/python2.7/site-packages/mock/mock.py", line 721, in __getattr__
    raise AttributeError(name)
AttributeError: assert_called_once
```

**Comment by ashcrow** [comment]
@mbarnes I can't seem to reproduce those errors. I tried with squashing and without (just in case there was a misplaced fixup) on clean source trees and I get no errors. Can you verify there is no locally modified code as well?

**Comment by mbarnes** [comment]
Did `pip install -U -r requirements.txt -r test-requirements.txt` and now the failures are magically gone.  Wheee!!

**Comment by mbarnes** [comment]
Merged manually since Travis is confused today.
Re-ran all the tests first because *I'm* confused today.

**Pull request closed without merging**